core/services/functions/orm.go

```diff
@@ -186,11 +189,47 @@ func (o orm) FindOldestEntriesByState(state RequestState, limit uint32, qopts ..
 	return requests, nil
 }
 
-func (o orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
+func (o *orm) FindById(requestID RequestID, qopts ...pg.QOpt) (*Request, error) {
 	var request Request
 	stmt := fmt.Sprintf(`SELECT %s FROM ocr2dr_requests WHERE request_id=$1 AND contract_address=$2;`, requestFields)
 	if err := o.q.WithOpts(qopts...).Get(&request, stmt, requestID, o.contractAddress); err != nil {
 		return nil, err
 	}
 	return &request, nil
 }
+
+func (o *orm) PruneOldestRequests(maxStoredRequests uint32, batchSize uint32, qopts ...pg.QOpt) (total uint32, pruned uint32, err error) {
+	err = o.q.WithOpts(qopts...).Transaction(func(tx pg.Queryer) error {
+		if err2 := tx.Get(&total, "SELECT COUNT(*) FROM ocr2dr_requests WHERE contract_address=$1", o.contractAddress); err2 != nil {
```

Comment:
I can't get rid of a feeling this is overkill to execute a count query.
Can't we make it time-based prune? E.g. delete oldest requests having received_at < deadline?

Reply:
We could but then we will either need to set a pretty short timeout or risk overloading the DB if we have a sudden influx of requests (spam, tooling bug,...).

Task run cleanup logic is also based on number of entries, see: https://github.com/smartcontractkit/chainlink/blob/develop/core/services/pipeline/orm.go#L642

Do you have an idea of how much faster the time-based approach will be in practice? I set defaults to 30k entires and a check every 5mins. I could adjust those.